### PR TITLE
JupyterLab WIPP extension 0.3.0

### PR DESCRIPTION
**Pull request by ktaletsk**
Next version of JupyterLab WIPP extension brings improvement to WIPP Collections sidebar:
- Support of multiple WIPP Collection types (Image Collections, CSV Collections)
- Search bar for WIPP Collection of chosen type
- Sorting of Collections by their attributes
- UI improvements

**Technical details:**
Table component was reimplemented in React for interactive sorting and updates. Component is generalized to take generic ImageCollection inputs, so it will be trivial to add new WIPP Collection types if needed.

**Screenshots:**

image

image

### CODE REVIEW

File: jupyterlab_wipp/src/sidebar.tsx [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+        updateWidget("");
+        this._searchBar.value = "";
+      }
+    });
+    layout.addWidget(clearButton);
+
+    // Search button
+    const searchButton = new ToolbarButton({
+      iconClassName: 'wipp-SearchIcon jp-Icon jp-Icon-16',
+      // onClick: searchFunction,
```
Comment (sunnielyu) [inline jupyterlab_wipp/src/sidebar.tsx]: Remove commented code

File: jupyterlab_wipp/jupyterlab_wipp/wipp.py [hunk 2]
```diff
@@ -99,7 +100,66 @@ def get_image_collections_all_pages(self):
 
     def get_image_collections(self):
         return [collection.json for collection in sum(self.get_image_collections_all_pages(), [])]
+
+    def search_image_collections_summary(self, name):
```
Comment (sunnielyu) [inline jupyterlab_wipp/jupyterlab_wipp/wipp.py]: Can you add comments before each function with a brief summary and explaining the expected inputs and outputs?

File: jupyterlab_wipp/jupyterlab_wipp/wipp.py [hunk 3]
```diff
@@ -99,7 +100,66 @@ def get_image_collections_all_pages(self):
...
+        r = requests.get(os.path.join(self.api_route, f'imagesCollections/search/findByNameContainingIgnoreCase?name={name}'))
```
Comment (sunnielyu) [inline jupyterlab_wipp/jupyterlab_wipp/wipp.py]: Can we add error handling to all the APi calls?

File: jupyterlab_wipp/jupyterlab_wipp/handlers.py [hunk 4]
```diff
@@ -49,14 +50,55 @@ def get(self):
...
+            }
+        """
+        data = json.loads(self.request.body.decode("utf-8"))
+        response = self.wipp.search_image_collections(data["name"])
+        
+        self.finish(json.dumps(response))
+
+class WippCsvCollections(WippHandler):
+    def get(self):
+        """
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_wipp/jupyterlab_wipp/handlers.py]: Fixed that

File: jupyterlab_wipp/jupyterlab_wipp/wipp.py (hunk 2 above)
Comment (ktaletsk) [reply jupyterlab_wipp/jupyterlab_wipp/wipp.py]: Done

File: jupyterlab_wipp/src/sidebar.tsx [hunk 5]
```diff
@@ -125,55 +124,56 @@ export class WippSidebar extends Widget {
...
-            insertInputPath(editor, "'/opt/shared/wipp/collections/" + id + "/images/'");
-          };
-        }
+      var app = this.app;
+      var notebookTracker = this.notebookTracker;
+      var consoleTracker = this.consoleTracker;
 
-        // specify table header
-        this._tableHeader = [['name', 'Name'], ['numberOfImages', '# of images'], ['imagesTotalSize', 'Total size'], ['creationDate', 'Creation date']];
+      var file_path_prefix: string;
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_wipp/src/sidebar.tsx]: done

File: jupyterlab_wipp/src/sidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+          </button>
+      </td>
+      
+    </tr>
+  )
+}
+  
+  
+// Generic class for different types of tables (ImageCollection, CsvCollection, etc)
+export class GenericTableWidget<T> extends Component<IGenericTableProps<IGenericCollection>, IGenericTableState<IGenericCollection>> {
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_wipp/src/sidebar.tsx]: done

File: jupyterlab_wipp/src/sidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+      if (response.status !== 200) {
+        return response.json().then(data => {
+          throw new ServerConnection.ResponseError(response, data.message);
+        });
+      }
+      return response.json();
+    });
+  }
+
+export interface ITableSignal {
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_wipp/src/sidebar.tsx]: Moved to `types.ts`

File: jupyterlab_wipp/jupyterlab_wipp/wipp.py (hunk 3 above)
Comment (ktaletsk) [reply jupyterlab_wipp/jupyterlab_wipp/wipp.py]: done

**Pull request merged**